### ISSUE TITLE
IR pipeline with empty optimizer step sequence generates wrong library link reference

### ISSUE THREAD

**Issue by fvictorio** [ethereum/solidity#13786]
## Description

Compiling this contract with `solc --via-ir --optimize --yul-optimizations "" --bin Foo.sol` generates a bytecode with a library reference:

```solidity
library Lib {
    function libFunction() internal pure returns (string memory) {
      return "";
    }
}

contract Foo {
  function f() public {
      Lib.libFunction();
      Lib.libFunction();
  }
}
```

The same contract compiled without IR, or with IR with the default steps, or with IR with `":"` as the steps, doesn't generate that reference.

Also: if the library function is called only once, this doesn't happen.

Might be related to https://github.com/ethereum/solidity/issues/12865?

## Environment

- Compiler version: Nightly (0.8.18-ci.2022.12.2+commit.591df042.Linux.clang)
- Target EVM version (as per compiler settings): Default
- Framework/IDE (e.g. Truffle or Remix): Solc CLI
- EVM execution environment / backend / blockchain client: -
- Operating system: Linux

**Comment by hrkrshnn** [comment]
This is a known issue that should be resolved when #12865 is fixed. Maybe we should really stop adding `linkersymbols` for internal library functions.

**Comment by ekpyron** [comment]
The weird thing here is that if ``""`` has the linker reference, then ``":"`` should have it even more so - would be worthwhile to figure out why that is apparently not the case. But yeah, that's rather to check if the optimizer sequence settings work as intended - the actual issue is indeed a duplicate of #12865, which we can easily and probably will do by means of https://github.com/ethereum/solidity/issues/12533#issuecomment-1298933401 - i.e. by enabling a minimal, non-invasive set of optimizations by default, which will get rid of these link references in Yul.

**Comment by nikola-matic** [comment]
Heya @fvictorio. So, I've done a bit of research  using the code snippet you've provided, and it turns out that this doesn't have much to do with the Yul optimizer - rather, the library reference gets optimized away by the libevmasm optimizer *after* the optimized IR has been generated. There is a slight difference with how the optimized IR looks between `""` and `":"` optimization steps, however, both maintain at least on `linkersymbol` call, which will then get propagated to the EVM assembly representation as `linkerSymbol`, and in the case of an empty optimization sequence, it doesn't get optimized away.

This is related to https://github.com/ethereum/solidity/issues/12865 as both yourself and @hrkrshnn noticed.

I'm closing this as we already have an issue open that will solve this. If you have any more questions, feel free to ask :)
